### PR TITLE
Argument should distinguish Input and Output Buffers

### PR DESCRIPTION
**Pull request by steven-johnson**

### CODE REVIEW

File: src/Argument.h [hunk 1]
```diff
@@ -28,19 +28,20 @@ struct Argument {
...
-     * If kind == Buffer, then type.bytes() should be used to determine
-     * elem_size of the buffer; additionally, type.code *should* reflect
-     * the expected interpretation of the buffer data (e.g. float vs int),
+     * If kind == InputBuffer|OutputBuffer, then type.bytes() should be used
+     * to determine* elem_size of the buffer; additionally, type.code *should*
+     * reflect the expected interpretation of the buffer data (e.g. float vs int),
      * but there is no runtime enforcement of this at present.
      */
     enum Kind {
         Scalar,
```
Comment (kree-colemcalughlin) [inline src/Argument.h]: Do we want to replace Scalar with InputScalar, OutputScalar now? Or can we provide the option of output scalars another way?

Comment (abadams) [inline src/Argument.h]: We could add these if we ever support input and output scalars. But given that we don't currently support them, best to leave it as is IMO.

File: src/CodeGen_GPU_Host.cpp [hunk 2]
```diff
@@ -188,9 +188,13 @@ vector<GPU_Argument> GPU_Host_Closure::arguments() {
         if (iter->second.write) debug(2) << " (write)";
         debug(2) << "\n";
 
-        GPU_Argument arg(iter->first, Argument::Buffer, iter->second.type, iter->second.dimensions, iter->second.size);
-        arg.read = iter->second.read;
-        arg.write = iter->second.write;
+        // TODO: read+write textures were already prohibited in the GLSL backend,
```
Comment (abadams) [inline src/CodeGen_GPU_Host.cpp]: I believe that no GPU backends use input/output buffers, so this should be OK.

File: src/CodeGen_OpenCL_Dev.cpp [hunk 3]
```diff
@@ -452,7 +452,7 @@ void CodeGen_OpenCL_Dev::CodeGen_OpenCL_C::add_kernel(Stmt s,
     for (size_t i = 0; i < args.size(); i++) {
         if (args[i].is_buffer()) {
             stream << " " << get_memory_space(args[i].name) << " ";
-            if (!args[i].write) stream << "const ";
+            if (!args[i].is_output()) stream << "const ";
```
Comment (abadams) [inline src/CodeGen_OpenCL_Dev.cpp]: Looks like a search-replace. I think if (args[i].is_input()) stream << "const " would be better.

File: test/correctness/constraints.cpp [hunk 4]
```diff
@@ -62,7 +62,7 @@ int main(int argc, char **argv) {
     error_occurred = false;
     h.realize(image1);
     // Also check it compiles ok without an inferred argument list
-    h.compile_to_assembly("h.s", Internal::vec<Argument>(image1), "h");
+    h.compile_to_assembly("h.s", Internal::vec<Argument>(Buffer(image1).as_input()), "h");
```
Comment (abadams) [inline test/correctness/constraints.cpp]: This is a subtle language change. Using an Image as an Argument for AOT compilation treats it as an ImageParam. Not including it in the Argument list makes it an embedded const image. This change forbids the former usage (without jumping through extra syntactic hoops).

File: src/CodeGen_OpenCL_Dev.cpp (hunk 3 above)
Comment (steven-johnson) [inline src/CodeGen_OpenCL_Dev.cpp]: Done

File: test/correctness/constraints.cpp (hunk 4 above)
Comment (steven-johnson) [inline test/correctness/constraints.cpp]: OK, I'll back out the change to Buffer, and have operator Argument always imply an InputBuffer.

**Comment by dsharletg** [comment]
I don't think we should prohibit input/output buffers for GPU backends. There's no reason it wouldn't work for OpenCL/CUDA, and in fact I would expect it to happen if a Func with an update step was scheduled on the GPU.

The limitiation in OpenGL is OpenGL-specific: OpenGL uses textures, and textures cannot be read/write because the texture cache doesn't support that. But OpenCL and CUDA expose buffers which don't have caches (or the cache supports invalidation).

**Comment by steven-johnson** [comment]
Ugh. In that case, having GPU_Argument be a subclass of Argument is just a really bad fit, since it implies possibilities that don't exist for "normal" Arguments.

**Comment by abadams** [comment]
Good point - cuda update steps probably use input/output buffers right now.
I retract my claim.

**Comment by steven-johnson** [comment]
Looking at this a little more, I think having GPU_Argument inherit from Argument is just a bad first in the first place. Let me push another yak onto the stack and investigate severing that relationship.

**Comment by dsharletg** [comment]
You might consider waiting for the modules branch for this change (https://github.com/halide/Halide/compare/modules). This is all going to change fairly dramatically with that branch (the compile method is very different), and it might make a lot of this easier. I'm going to submit it as a PR very shortly (it has a dependency on #560, which is going to be merged any time now).

If nothing else, you'll save me a merge headache :)

**Comment by steven-johnson** [comment]
> You might consider waiting for the modules branch

Is there an ETA? This seemingly-small thing has me blocked several layers up the stack.

**Comment by dsharletg** [comment]
I will submit a PR for it by EOD. I'm diagnosing new test failures right
now, I think they aren't too hard to fix.

BTW, if it helps, what I've been doing is just working in new branches
forked from another branch that is yet-to-be-merged. The advantage is it
lets you make progress right away, the disadvantage is it won't get merged
until the branch you forked from gets merged.

**Pull request closed without merging**